### PR TITLE
test: add failing test for object with constructor

### PR DESCRIPTION
**Pull request by Chaoste**
The following currently causes a fatal error with this module (while working with lodash):
```
copy({ 'sys[constructor]': 'something' })
```

The error looks like this:
```
TypeError: Function.prototype.toString required that 'this' be a Function
    at  String.toString.(<anonymous>)
    at  getCleanClone (/app/node_modules/fast-copy/src/utils.ts:76:25
    at  getObjectCloneLoose (/app/node_modules/fast-copy/src/utils.ts:103:22
    at  handleCopy (/app/node_modules/fast-copy/src/index.ts:182:12
    at  getObjectCloneLoose (/app/node_modules/fast-copy/src/utils.ts:109:20
    at  handleCopy (/app/node_modules/fast-copy/src/index.ts:72:14
    at  copy (/app/node_modules/fast-copy/src/index.ts:185:10
```

Version used: `"fast-copy": "^2.1.1"`

I added a test case to confirm that this is an error. Once I got a failing test as confirmation, I would implement the first solution from [this SO post](https://stackoverflow.com/a/19717946/17269164) which seems to be quite performant.

Greetings,
Thomas

cc @johanneswuerbach

### CODE REVIEW

**Comment by Chaoste** [comment]
@planttheidea Hey I just provided this fix and would be happy for a quick review.

**Comment by Chaoste** [comment]
@planttheidea Hey, would be great if you review this PR

**Comment by planttheidea** [comment]
> @planttheidea Hey, would be great if you review this PR

Apologies for the delay; my wife just had our second child last week. This is in my queue, and was planning on tackling it sometime this week.

**Comment by Chaoste** [comment]
> > @planttheidea Hey, would be great if you review this PR
> 
> Apologies for the delay; my wife just had our second child last week. This is in my queue, and was planning on tackling it sometime this week.

Wow, Gratulations! I guess that's a reasonable excuse. Next week would be fine as well ;)

File: src/utils.ts
Comment (planttheidea) [inline src/utils.ts:65]: Two things:

While I think this will work around the problem, it is not a true "copy". In your test example, it is copying a standard object, which would have a prototype of `Object`, but the copy would have a prototype of `null`, meaning all the standard prototype methods would not be inherited. This is also true of custom prototype objects, because JS can be super flexible:

```js
const instance = new CustomThing();

instance.constructor = 'hello';
```

I would instead recommend gating the `toString` check on the fact that `constructor` is a function:

```ts
export const getCleanClone = (object: any, realm: FastCopy.Realm): any => {
  if (!object.constructor) {
    return create(null);
  }

  const { constructor: Constructor } = object;
  const prototype = object.__proto__ || getPrototypeOf(object);

  if (Constructor === realm.Object) {
    return prototype === realm.Object.prototype ? {} : create(prototype);
  }

  if (typeof Constructor === 'function' && ~toStringFunction.call(Constructor).indexOf('[native code]')) {
    try {
      return new Constructor();
    } catch {}
  }

  return create(prototype);
};
```

This would eliminate the error, but also ensure the custom object was created with the correct prototype.

File: __tests__/index.ts
Comment (planttheidea) [inline __tests__/index.ts:282]: In the vein of my other comment, I would recommend checking that the prototypes of the original and copy are the same. I would also create a similar test, but with a custom object instead of a standard POJO.

**Review by planttheidea — CHANGES_REQUESTED** [review]

File: src/utils.ts
Comment (planttheidea) [reply src/utils.ts:65]: One additional thing: noticed I used `typeof` instead of `instanceof`. In modern engines the latter is quite fast, however in older engines it is significantly slower. For this purpose there would be no result differences between the two, so we should default to the one that is fast in all engines.

File: __tests__/index.ts
Comment (Chaoste) [reply __tests__/index.ts]: Added a test case with a new class `Bar` that holds a string at `constructor`. I started by writing the second test to see that my current approach fails and then fixed it with your suggestion.

**Review by planttheidea — APPROVED** [review]
LGTM after the changes, and appreciate the increased test coverage. I'll wait for you to merge it (whenever you are ready), and once merged I'll try to carve out a little time to publish the patch. Thanks for the contribution!

**Comment by Chaoste** [comment]
Awesome! I'm ready to merge but don't have write access. Feel free to merge it :)

**Comment by planttheidea** [comment]
FYI I've identified a few other gaps that this can occur with that are extreme edge cases (example: the exact same `constructor` override on an `Array`). I'm working on a change to accommodate this in a more universal way, but should have the change in by today.

**Comment by planttheidea** [comment]
Alright, was able to get things sorted, as well as make some long-standing cleanup enhancements. I have published `2.1.2` with this change, let me know if you have any issues.

Thanks for the contribution!

**Pull request merged**
